codec_packetizers/h265/h265_depacketizer.c

```diff
@@ -0,0 +1,505 @@
+                result = H265_RESULT_OUT_OF_MEMORY;
+                break;
+            }
+            curNaluDataIndex += NALU_HEADER_SIZE;
+        }
+
+/* Write NALU payload */
+        size_t payloadLength = curPacketLength - TOTAL_FU_HEADER_SIZE;
+        if( ( curNaluDataIndex + payloadLength ) <= pNalu->naluDataLength )
+        {
+            memcpy( &( pNalu->pNaluData[curNaluDataIndex] ),
+                    &( pCurPacketData[TOTAL_FU_HEADER_SIZE] ),
+                    payloadLength );
+        }
+        else
+        {
+            result = H265_RESULT_OUT_OF_MEMORY;
+            break;
+        }
+        curNaluDataIndex += payloadLength;
+
+/* Move to next packet */
+        pCtx->tailIndex++;
+        pCtx->packetCount--;
+    }
+
+/* Update final NALU length */
+    pNalu->naluDataLength = curNaluDataIndex;
+
+    return result;
+}
+
+
+
+
+/*-------------------------------------------------------------------------------------------------------------*/
+
+static H265Result_t H265Depacketizer_ProcessAggregationPacket( H265DepacketizerContext_t * pCtx,
+                                                               H265Nalu_t * pNalu )
+{
+    H265Result_t result = H265_RESULT_OK;
+    H265Packet_t * pCurrentPacket;
+    size_t minSize;
+    size_t offset;
+    uint16_t nalu_size;
+
+/* Get current packet */
+    pCurrentPacket = &pCtx->pPacketsArray[pCtx->tailIndex];
+
+/* Calculate minimum required size */
+    minSize = AP_HEADER_SIZE + ( 2 * AP_NALU_LENGTH_FIELD_SIZE + 2 * NALU_HEADER_SIZE );
+
+    if( pCurrentPacket->packetDataLength > minSize )
+    {
+/* If this is first NAL unit in AP, initialize state */
+        if( pCtx->currentlyProcessingPacket != H265_AP_PACKET )
+        {
+            pCtx->currentlyProcessingPacket = H265_AP_PACKET;
+            pCtx->apDepacketizationState.currentOffset = AP_HEADER_SIZE;
+            pCtx->apDepacketizationState.firstUnit = 1;
+
+/* Get first NAL unit size */
+            offset = pCtx->apDepacketizationState.currentOffset;
+            nalu_size = ( pCurrentPacket->pPacketData[offset] << 8 ) |
+                        pCurrentPacket->pPacketData[offset + 1];
+            offset += AP_NALU_LENGTH_FIELD_SIZE;
+
+/* Validate NALU size */
+            if( offset + nalu_size <= pCurrentPacket->packetDataLength )
+            {
+/* Add buffer size check */
+                if( nalu_size <= pNalu->naluDataLength )
+                {
+/* Direct memcpy for first unit */
+                    memcpy( pNalu->pNaluData,
+                            &pCurrentPacket->pPacketData[offset],
+                            nalu_size );
+
+/* Update NALU length */
+                    pNalu->naluDataLength = nalu_size;
+
+/* Update offset for next time */
+                    offset += nalu_size;
+                    pCtx->apDepacketizationState.currentOffset = offset;
+                }
+                else
+                {
+                    result = H265_RESULT_OUT_OF_MEMORY;
```

Comment:
Should we drop current AP data when it fail to depacketize it?
Do we need to clear these variables?
- currentlyProcessingPacket
- apDepacketizationState.*

Do we need to update below variables to skip this failing packet?
- tailIndex
- packetCount

Reply:
This question applies to all failing results in this function.